test/src/test/integration/git_protocol.rs

```diff
@@ -0,0 +1,384 @@
+            ref_prefixes: None,
+            want_refs: vec![],
+            done_after_pack: true,
+        },
+        want_id,
+        |_| packwriter::Discard,
+    )
+    .unwrap();
+
+    assert!(server_out.success());
+    assert!(client_out.pack.is_none());
+}
+
+fn clone_with<R, L, B, P>(remote: R, local: L, build_pack_writer: B)
+where
+    R: AsRef<Path>,
+    L: AsRef<Path>,
+    B: FnOnce(Arc<AtomicBool>) -> P,
+    P: PackWriter + Send + 'static,
+    P::Output: Send + 'static,
+{
+    let (client_out, server_out) = run_fetch(
+        &remote,
+        fetch::Options {
+            repo: "foo".into(),
+            extra_params: vec![],
+            ref_prefixes: Some(vec![]),
+            want_refs: vec![],
+            done_after_pack: true,
+        },
+        want_id,
+        build_pack_writer,
+    )
+    .unwrap();
+
+    assert!(server_out.success());
+    assert!(client_out.pack.is_some());
+
+    let remote_repo = git2::Repository::open(remote).unwrap();
+    remote_repo.set_namespace("foo").unwrap();
+    let local_repo = git2::Repository::open(&local).unwrap();
+
+    update_tips(&local_repo, &client_out.refs).unwrap();
+
+    let mut remote_refs = collect_refs(&remote_repo).unwrap();
+    let mut local_refs = collect_refs(&local_repo).unwrap();
+
+    remote_refs.sort();
+    local_refs.sort();
+
+    assert_eq!(remote_refs, local_refs);
+}
+
+#[test]
+fn clone_libgit() {
+    let remote = upstream();
+    let local = tempdir().unwrap();
+    let local_repo = git2::Repository::init(&local).unwrap();
+
+    clone_with(&remote, &local, move |stop| {
+        packwriter::Libgit::new(packwriter::Options::default(), local_repo, stop)
+    })
+}
+
+#[test]
+fn clone_gitoxide() {
+    let remote = upstream();
+    let local = tempdir().unwrap();
+    let local_repo = git2::Repository::init(&local).unwrap();
+
+    clone_with(&remote, &local, move |stop| {
+        packwriter::Standard::new(local_repo.path(), packwriter::Options::default(), stop)
+    })
+}
+
+fn thin_pack_with<R, L, B, P>(remote: R, local: L, build_pack_writer: B)
+where
+    R: AsRef<Path>,
+    L: AsRef<Path>,
+    B: Fn(Arc<AtomicBool>) -> P,
+    P: PackWriter + Send + 'static,
+    P::Output: Send + 'static,
+{
+    // Clone main only
+    {
+        let (client_out, server_out) = run_fetch(
+            &remote,
+            fetch::Options {
+                repo: "foo".into(),
+                extra_params: vec![],
+                ref_prefixes: None,
+                want_refs: vec![WantRef {
+                    name: "refs/heads/main".into(),
+                    have: None,
+                }],
+                done_after_pack: true,
+            },
+            want_id,
+            &build_pack_writer,
+        )
+        .unwrap();
+
+        assert!(server_out.success());
+        assert!(client_out.pack.is_some());
+    }
+
+    let remote_repo = git2::Repository::open(&remote).unwrap();
+    remote_repo.set_namespace("foo").unwrap();
```

Comment:
Adding namespace support is on the list of features to add to the gitoxide loose ref store now.